### PR TITLE
[mlir][tensor] Fold padding_value away for pack ops when possible.

### LINKED ISSUE #15417: Fold padding_value away if a tensor.pack op does not have incomplete tile
Found the case in IR dump and bug reports. We want the folder, which folds padding_value away if possible. It can be applied when we have enough static information. E.g, 500000 is divisible by 16 and 1200 is divisible by 1. In this context, we don't need padding_value.

```mlir
module {
  func.func @main_dispatch_1297() {
    %c0 = arith.constant 0 : index
    %c12480 = arith.constant 12480 : index
    %cst = arith.constant 0.000000e+00 : f32
    %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<1200x500000xf32>>
    %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c12480) : !flow.dispatch.tensor<writeonly:tensor<31250x1200x16x1xf32>>
    %2 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [1200, 500000], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<1200x500000xf32>> -> tensor<1200x500000xf32>
    %3 = tensor.empty() : tensor<31250x1200x16x1xf32>
    %pack = tensor.pack %2 padding_value(%cst : f32) outer_dims_perm = [1, 0] inner_dims_pos = [1, 0] inner_tiles = [16, 1] into %3 : tensor<1200x500000xf32> -> tensor<31250x1200x16x1xf32>
    flow.dispatch.tensor.store %pack, %1, offsets = [0, 0, 0, 0], sizes = [31250, 1200, 16, 1], strides = [1, 1, 1, 1] : tensor<31250x1200x16x1xf32> -> !flow.dispatch.tensor<writeonly:tensor<31250x1200x16x1xf32>>
    return
  }
}
```

### PR DESCRIPTION
**Pull request by hanhanW**
If we can infer statically that there are no incomplete tiles, we can remove the optional padding operand.

Fixes https://github.com/openxla/iree/issues/15417

### CODE REVIEW

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 1]
```diff
@@ -3800,17 +3800,37 @@ static bool haveSameTiles(PackOp packOp, UnPackOp unPackOp) {
   return true;
 }
 
-/// Fold an unpack(pack(x)) to x.
+/// Returns true if the pack op does not need a padding value.
+static bool paddingIsNotNeeded(PackOp op) {
+  auto srcType = op.getSourceType();
+  if (llvm::any_of(op.getInnerDimsPos(),
```
Comment (chelini) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:3806]: What if the tiles are not statically known? It looks like we continue in `requirePaddingValue`; Can we also add a test for the tiles?

Comment (hanhanW) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:3806]: This is why we check if those dim sizes are dynamic or not in the first beginning. I think we can only do nothing if they are dynamic tiles.

Comment (hanhanW) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:3806]: oh you meant dynamic inner tile sizes? Good catch, adding the case now...

**Review by chelini — APPROVED** [review]
thanks, looks good to me.

**Pull request merged**